### ISSUE TITLE
What is `KCFG.Cover.csubst` supposed to be?

### ISSUE THREAD

**Issue by Stevengre** [runtimeverification/k#4602]
I suspect that `KCFG.Cover.csubst` should be able to obtain the `source` by applying it to the `target`. However, in its current form, the `match_with_constraint` in `create_cover` doesn’t achieve this when it’s set to `None`.

https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/kcfg/kcfg.py#L854-L863

There is a non-None call here, which should have the same effect as a non-empty call. However, based on [this documentation](https://github.com/runtimeverification/haskell-backend/blob/2faf6c1d71de054f4630e3489efb4631a0f320b5/docs/2022-07-18-JSON-RPC-Server-API.md?plain=1#L361), it doesn’t seem to behave the same way.

https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/proof/reachability.py#L193-L194

So what exactly does it do? How can we make it consistent? And do we even need `csubst` if `match_with_constraint` works correctly?

**Comment by tothtamas28** [comment]
The only call to `create_cover` with `csubst=None`: https://github.com/runtimeverification/k/blob/eb0b1815af0a8c08773ce43c70f9f307624d195f/pyk/src/pyk/kcfg/kcfg.py#L530-L533

**Comment by tothtamas28** [comment]
TL;DR: When the `APRPRover` establishes coverage between nodes, the `KCFG.Cover.csubst.constraints` are the conjuncts of the `predicate` from the [KORE RPC `implies` response](https://github.com/runtimeverification/haskell-backend/blob/master/docs/2022-07-18-JSON-RPC-Server-API.md#correct-response-1).

---

The other code path:

`KCFG.Cover` is instantiated in `APRProof.commit`: https://github.com/runtimeverification/k/blob/eb0b1815af0a8c08773ce43c70f9f307624d195f/pyk/src/pyk/proof/reachability.py#L234-L235

`APRProofSubsumeResult` is instantiated in `APRProver.step_proof`: https://github.com/runtimeverification/k/blob/eb0b1815af0a8c08773ce43c70f9f307624d195f/pyk/src/pyk/proof/reachability.py#L811-L820

There, `csubsts` is computed in `APRProver._check_subsume`: https://github.com/runtimeverification/k/blob/eb0b1815af0a8c08773ce43c70f9f307624d195f/pyk/src/pyk/proof/reachability.py#L782-L786

which calls `CTermSymbolic.implies`: https://github.com/runtimeverification/k/blob/eb0b1815af0a8c08773ce43c70f9f307624d195f/pyk/src/pyk/cterm/symbolic.py#L215-L217

where the constraints are populated from the response `predicate`: https://github.com/runtimeverification/k/blob/eb0b1815af0a8c08773ce43c70f9f307624d195f/pyk/src/pyk/cterm/symbolic.py#L269-L270

https://github.com/runtimeverification/k/blob/eb0b1815af0a8c08773ce43c70f9f307624d195f/pyk/src/pyk/cterm/symbolic.py#L272-L273

https://github.com/runtimeverification/k/blob/eb0b1815af0a8c08773ce43c70f9f307624d195f/pyk/src/pyk/cterm/symbolic.py#L282-L283

KORE RPC docs: https://github.com/runtimeverification/haskell-backend/blob/master/docs/2022-07-18-JSON-RPC-Server-API.md#correct-response-1

**Comment by Stevengre** [comment]
Can we close this issue, because we've decided how the `CSubst.apply` performs. @tothtamas28
